### PR TITLE
Improved logging system

### PR DESCRIPTION
**Pull request by MikeMatrix**
Improved logging system with a more flexible approach to handling when to log, how the log looks and what to log.
This can be used in pretty much any situation and can be easily expanded by either internal code or third party code.

The intention of this is to standardize logging not only for RPT logs, make it readable and add enough metadata to be useful in any situation.

As it stands right now, the code allows you to dynamically hook log writers, that will actually perform the output. This means you can easily create a logger that outputs a formatted log to the RPT as well as systemChat, on every client's RPT, etc.
At this point it only allows you to put copies of the function in, which will be expanded into being able to reference functions with Strings, so you can call them directly instead of passing in a copy of them.

I feel the logging format right now seems acceptable, but it is still being worked on and tweaked for better readability.

Still WIP.

### CODE REVIEW

File: addons/log/config.cpp [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+class CfgPatches {
+    class ADDON {
+        units[] = {};
+        weapons[] = {};
+        requiredVersion = REQUIRED_VERSION;
+        requiredAddons[] = { "cba_common" };
+        version = VERSION;
+        author[] = { "MikeMatrix" };
+        authorUrl = "http://dev-heaven.net/projects/cca";
```
Comment (Killswitch00) [inline addons/log/config.cpp]: The project URL is https://github.com/CBATeam/CBA_A3 nowadays.

**Comment by MikeMatrix** [comment]
Will change that with the next changes. Still have some improvements to do.
I'll probably add a log level stack instead, so you can temporary force a different log level for a specific part without losing the previous log level.

File: addons/log/config.cpp (hunk 1 above)
Comment (MikeMatrix) [inline addons/log/config.cpp]: Will change that with the next changes.

**Comment by PabstMirror** [comment]
Problem with this kind of dynamic logging is it turns macros that would normally be compiled out into function calls.

`TRACE_1("test",_keys)`
becomes 
```
["cba", "hashes", format ['%1:  _keys=%2', "test", if (isNil{ _keys}) then {nil} else { _keys}], 10, ["CBA_fnc_diagLogWriter"], "x\cba\addons\hashes\fnc_hashCreate.sqf", 42] call CBA_fnc_logDynamic;
```
instead of just becoming `/* disabled */`- and having no effect.

We gain the advantage of allowing someone to debug, without needing to have raw files, or mess with filepatching; but there is going to be a performance cost.  We can still add dynamic debuing, but I think they need to be on new macros and not modify the old ones.

**Pull request closed without merging**